Adding gfx9 pc sampling (#5283)
ROCm/rocm-systems: projects/rocprofiler-sdk/source/docs/how-to/cdna3-cdna4-pc-sampling.rst

.. meta::
  :description: Analyzing PC sampling data on gfx9 architecture
  :keywords: Sampling PC, Sampling program counter, rocprofv3, rocprofv3 tool usage, Using rocprofv3, ROCprofiler-SDK command line tool, PC sampling

.. _gfx9-pc-sampling:

> 

Reviewer: The name of this RST file should better align with the title of the page.

In addition to the wave's state, PC sampling captures the state of the sampled SIMD's arbiter (also referred to as the `scheduler <https://rocm.docs.amd.com/projects/rocprofiler-compute/en/latest/conceptual/pipeline-descriptions.html#scheduler>`_). The term arbiter is used in accordance with the PC sampling data fields. The arbiter state indicates whether any wavefront on the SIMD was issued to a given pipeline during the given cycle, and if so, whether that instruction began execution.

PC sampling on the CDNA3 (AMD Instinct MI300 Series) and CDNA4 (AMD Instinct MI350 Series) architectures primarily focuses on the frontend of shader execution, examining which waves are running and what prevents them from issuing instructions. It provides a limited view of the backend (execution pipelines), examining whether they are stalled and back-pressuring the front-end, but not the underlying cause of the stall. For the list of pipelines available on the CDNA3 and CDNA4 architecture, see :ref:`execution-pipelines`.

Stall reasons
==============

> The stall reason reported for a sample indicates why the sampled wave couldn't issue or execute an instruction during the given cycle. This contrasts with the arbiter state fields, which report the entire SIMD state. To learn more about arbiter state, see :ref:`arbiter_state`.

Reviewer 2: ```suggestion
The stall reason reported for a sample indicates why the sampled wave couldn't issue or execute an instruction during the given cycle. This contrasts with the arbiter state fields, which report the entire SIMD state. To learn more about the arbiter state, see :ref:`arbiter_state`.
```

The following table lists the stall reasons:

.. list-table:: Stall reasons
  :header-rows: 1

  * - Stall reason
    - Description

  * - NO_INSTRUCTION_AVAILABLE
    - The wave is stalled waiting for instructions. For example, at the branch target, I$ miss, and others.

  * - ALU_DEPENDENCY
    - The sampled wave's instruction couldn't be issued due to an internal hardware dependency, such as an inter-pipeline dependency or a data hazard. For more information, see 4.4 Data dependency resolution in `AMD Instinct MI300 (CDNA3) instruction set architecture <https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/amd-instinct-mi300-cdna3-instruction-set-architecture.pdf>`_ or `CDNA4 instruction set architecture <https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/amd-instinct-cdna4-instruction-set-architecture.pdf>`_.

  * - WAITCNT
    - The sampled wave is waiting due to memory dependency (``waitcnt``).

  * - INTERNAL_INSTRUCTION

>     - The sampled wave is issuing an internal instruction such as a ``NOP``.

Reviewer 2: ```suggestion
    - The sampled wave is issuing an internal instruction, such as a ``NOP``.
```

  * - BARRIER_WAIT
    - The sampled wave is waiting at a barrier for the other waves in the workgroup to reach that barrier.

  * - ARBITER_NOT_WIN
    - The sampled wave isn't selected to issue instructions. This typically occurs when multiple waves compete to issue instructions of the same type (served by the same execution pipeline) simultaneously. Only one wave wins the arbitration. For more information about wave arbitration, see `Scheduler documentation <https://rocm.docs.amd.com/projects/rocprofiler-compute/en/latest/conceptual/pipeline-descriptions.html#scheduler>`_.

  * - ARBITER_WIN_EX_STALL

>     - The wave was ready to issue an instruction, which the arbiter then selected to issue. However, the execution pipeline backpressured the wave, preventing it from issuing instructions because it couldn't accept more instructions.

Reviewer 3: ```suggestion
    - The wave was ready to issue an instruction, which the arbiter then selected to issue. However, the execution pipeline backpressured the wave, preventing it from issuing instruction because it couldn't accept more instructions.
```

Reviewer 3: @SwRaw are you planning to address this comment?

  * - OTHER

>     - Other reasons for stalling a wave such as recoverable page-fault (``XNACK``).

Reviewer 2: ```suggestion
    - Other reasons for stalling a wave, such as a recoverable page-fault (``XNACK``).
```

Occupancy
==========

> ``wave_cnt`` reports the number of active waves present on the CU (across all four SIMDs) at the time of sampling. This information helps explain the evolution of CU occupancy over time. The information includes details such as whether the CU was fully loaded throughout kernel execution, whether waves were draining towards the end, how occupancy affects the cost of stalls, and the occupancy at specific source or assembly lines at the time of sampling.

Reviewer 2: ```suggestion
``wave_cnt`` reports the number of active waves present on the CU (across all four SIMDs) at the time of sampling. This information helps explain the evolution of CU occupancy over time. The information includes details, such as whether the CU was fully loaded throughout kernel execution, whether waves were draining towards the end, how occupancy affects the cost of stalls, and the occupancy at specific source or assembly lines at the time of sampling.
```

.. list-table:: Diagnosing arbiter acitivity
  :header-rows: 1

  * - Condition
    - Diagnosis

  * - ``arbiter_state_issue_PIPE`` = true/false
    - ARBITER_NOT_WIN stalls. For details, see :ref:`arbiter_not_win`.

  * - ``arbiter_state_issue_PIPE`` == 0 OR ``arbiter_state_stall_PIPE`` == 1
    - | For one pipe, Pipe latency.
      | For all pipes, SIMD latency.
      | For details, see :ref:`latency_stalls`.

  * - ``arbiter_state_issue_PIPE`` == 0 AND ``arbiter_state_stall_PIPE`` == 0
    - Frontend pipe latency. For details, see :ref:`latency_stalls`.

  * - ``arbiter_state_issue_PIPE`` == 1 AND ``arbiter_state_stall_PIPE`` == 0
    - Instructions issued per cycle (IPC) estimation. For details, see :ref:`ipc_approximation`.

  * - ``arbiter_state_issue_PIPE`` == 1 AND ``arbiter_state_stall_PIPE`` == 0 with the wave stalled due to ``ARBITER NOT WIN``
    - Pipeline oversubscription. For details, see :ref:`pipeline_oversubscribe`.

  * - ``arbiter_state_stall_PIPE`` == 1 with the wave stalled due to ``ARBITER_WIN_EX_STALL``
    - Pipeline backpressuring. For details, see :ref:`pipeline-backpressure`.

.. _arbiter_not_win:

ARBITER_NOT_WIN stalls
-----------------------

> To diagnose the ARBITER_NOT_WIN stalls, check the ``arbiter_state_issue_PIPE`` value:
> 
> - ``arbiter_state_issue_PIPE`` = true: Another wave won the arbitration for that PIPE in the sampled cycle. This confirms the classic multiwave contention scenario. To determine whether the winning wave's instruction actually began execution, check ``arbiter_state_stall_PIPE``. Also, look out for pipeline hotspotting as high contention on a single pipe is not always beneficial.
> 
> - ``arbiter_state_issue_PIPE`` = false: No wave was issued on that PIPE in the sampled cycle. This indicates that the PIPE wasn't ready to accept an instruction in the sampled cycle. This is a rare scenario.
> 

Reviewer: ```suggestion
To diagnose the ARBITER_NOT_WIN stalls, check the ``arbiter_state_issue_PIPE`` value:

- ``arbiter_state_issue_PIPE`` = true: Another wave won the arbitration for that PIPE in the sampled cycle. This confirms the classic multiwave contention scenario. To determine whether the winning wave's instruction actually began execution, check ``arbiter_state_stall_PIPE``. Also, look for pipeline hotspotting, as high contention on a single pipe is not always beneficial.

- ``arbiter_state_issue_PIPE`` = false: No wave was issued on that PIPE in the sampled cycle. This indicates that the PIPE wasn't ready to accept an instruction in the sampled cycle and is a rare scenario.

```

------------------------------------

To identify pipeline oversubscription, look for samples where a wave is stalled due to ARBITER_NOT_WIN, despite the corresponding pipe having simultaneously accepted an instruction (indicated by ``arbiter_state_issue_PIPE`` == 1 AND ``arbiter_state_stall_PIPE`` == 0). This indicates that multiple waves on the sampled SIMD were competing to issue on the same pipe, and the pipe was able to serve at least one of those requests.

.. _pipeline-backpressure:

Pipeline backpressuring
------------------------

Sometimes, the execution pipeline backpressures a wave from issuing an instruction, even when the arbiter has chosen that wave. This is indicated by ``arbiter_state_stall_PIPE`` == 1 .
A backpressured wave will be stalled due to ARBITER_WIN_EX_STALL.

Pipeline backpressuring can occur due to the following conditions:

- **Oversubscription:** When a pipeline is oversubscribed, it might start backpressuring waves. Some pipes can accept only a limited number of outstanding instructions at a time. When that limit is exceeded, the pipe stops accepting new instructions and backpressures waves trying to issue instructions.

> - **long-latency instructions:** Backpressuring can also occur without pipeline oversubscription. For example, waves issuing instructions targeting different pipelines tend to cause no contention. However, some instructions, such as MFMA operations and vector transcendentals, are long-latency and require multiple quad-cycles to execute. If a wave attempts to issue long-latency instructions back-to-back, the pipe might backpressure it.

Reviewer 2: ```suggestion
- **long-latency instructions:** Backpressuring can also occur without pipeline oversubscription. For example, waves issuing instructions targeting different pipelines do not tend to cause contention. However, some instructions, such as MFMA operations and vector transcendentals, are long latency and require multiple quad-cycles to execute. If a wave attempts to issue long-latency instructions back-to-back, the pipe might backpressure it.
```
